### PR TITLE
Boost Lora with validate/ensemble/1trigger, Add Inpaint in backend 

### PR DESCRIPTION
**Pull request by wuziheng**
This PR provide
1. A validate and faceid ensemble which will improve original lora training , and make inference better, as we see pic1.
2. Add 1trigger tirck for face tagging which will stablize training and simplify the online prompt engineering.
3. Build a Inpaint tab support use lora + Arbitrary template to generate artistic photos in the backend. **After discussing with the core team, we've decided to only add the code for now without displaying it on the Gradio.**
4. Other small tricks and optimization for prompts and preprocess

All images are for demonstration of code functionality only. If there are any copyright issues, please contact me for removal.
We believe this is a big version for FC that requires everyone to discuss together.

L1VzZXJzL3d6aC9MaWJyYXJ5L0FwcGxpY2F0aW9uIFN1cHBvcnQvaURpbmdUYWxrLzE3MjQxNzE5NF92Mi9JbWFnZUZpbGVzLzIxMDAxLzE2OTI5NDg3OTUxMTJf5oiq5bGPMjAyMy0wOC0yNSDkuIvljYgzLjMzLjEyLnBuZw==

截屏2023-08-26 上午8 43 34
截屏2023-08-26 上午8 43 27

### CODE REVIEW

**Comment by wuziheng** [comment]
截屏2023-08-23 下午3 05 21
一个基于新链路lora 和原有链路训练的lora 在inpaint 链路下搭配不同base model的记录

File: app.py [hunk 1]
```diff
@@ -75,6 +88,33 @@ def train_lora_fn(foundation_model_path=None, revision=None, output_img_dir=None
         f'--lora_r=32 --lora_alpha=32 --lora_text_encoder_r=32 --lora_text_encoder_alpha=32')
 
 
+def train_lora_fn_paiya(foundation_model_path=None, revision=None, output_img_dir=None, work_dir=None):
+    os.system(
+        f'''
+        accelerate launch --mixed_precision="fp16" facechain/train_text_to_image_paiya.py \
+            --pretrained_model_name_or_path="{foundation_model_path}" \
+            --model_cache_dir='/mnt/zhoulou.wzh/AIGC/model_data/' \
```
Comment (wenmengzhou) [inline app.py:96]: remove hardcode dir

File: facechain/train_text_to_image_paiya.py [hunk 2]
```diff
@@ -0,0 +1,1297 @@
...
+        if np.sum(np.array(mask)) != 0:
+            images.append(mask_sub_image)
+
+    # 写入结果
+    for index, base64_pilimage in enumerate(images):
+        image = base64_pilimage.convert("RGB")
+        image.save(os.path.join(images_save_path, str(index) + ".jpg"))
+        print("save processed image to "+ os.path.join(images_save_path, str(index) + ".jpg"))
+        with open(os.path.join(images_save_path, str(index) + ".txt"), "w") as f:
+            f.write("zhoumo_face, zhoumo, 1person")
```
Comment (wenmengzhou) [inline facechain/train_text_to_image_paiya.py:193]: maybe a better prompt？

File: facechain/kohya_lora.py [hunk 3]
```diff
@@ -0,0 +1,1310 @@
+# LoRA network module
```
Comment (wenmengzhou) [inline facechain/kohya_lora.py:1]: not used, detele

File: app.py [hunk 4]
```diff
@@ -261,7 +393,11 @@ def train_input():
 
         run_button = gr.Button('开始训练（等待上传图片加载显示出来再点，否则会报错）'
                                'Start training (please wait until photo(s) fully uploaded, otherwise it may result in training failure)')
-
+        
+        use_paiya = gr.Radio(
+            label="Enable Inpaint : False means only support Inference, True support Inference/Inpaint and pleas use > 5 images", type="value", choices=[True, False],
```
Comment (wenmengzhou) [inline app.py:398]: also add chinese description

File: facechain/train_text_to_image_paiya.py [hunk 5]
```diff
@@ -0,0 +1,1297 @@
+# coding=utf-8
```
Comment (wenmengzhou) [inline facechain/train_text_to_image_paiya.py:1]: train_text_to_image_inpaint？

File: facechain/train_text_to_image_paiya.py [hunk 6]
```diff
@@ -0,0 +1,1297 @@
...
+            score       = face_quality_func(sub_image)[OutputKeys.SCORES]
+            score       = 0 if score is None else score[0]
+
+            face_id_scores.append(embedding)
+            quality_scores.append(score)
+            face_angles.append(angle)
+
+            selected_paths.append(jpg)
+        except:
+            pass
```
Comment (wenmengzhou) [inline facechain/train_text_to_image_paiya.py:135]: face pick module should be extracted to be used by developer more conviniently

File: app.py (hunk 4 above)
Comment (wuziheng) [reply app.py]: delete now

File: app.py [hunk 7]
```diff
@@ -157,6 +184,67 @@ def launch_pipeline(uuid,
...
+
+    # we use ensemble model, if not exists fallback to original lora
+    lora_model_path = f'/tmp/{uuid}/{output_model_name}/ensemble/'
+    if not os.path.exists(lora_model_path):
+        lora_model_path = f'/tmp/{uuid}/{output_model_name}/'
+
+    gen_portrait_inpaint = GenPortraitInpaint(crop_template=False, short_side_resize=512)
+    
+    # TODO this cache_model_dir & base_model_path should fix with snapshot_download
+    cache_model_dir = '/mnt/zhoulou.wzh/AIGC/model_data/'
```
Comment (yingdachen) [inline app.py:218]: debug代码先修正一下

File: app.py [hunk 8]
```diff
@@ -157,6 +184,67 @@ def launch_pipeline(uuid,
...
+    base_model_path = os.path.join('/mnt/workspace/.cache/modelscope/', base_model, 'film/film')
```
Comment (yingdachen) [inline app.py:219]: 同上

File: app.py [hunk 9]
```diff
@@ -289,62 +384,191 @@ def train_input():
...
+        display_button.click(
+            fn=launch_pipeline,
+            inputs=[uuid, pos_prompt, user_models, num_images, style_model, multiplier_style],
+            outputs=[infer_progress, output_images]
+        )
+        
     return demo
 
+# Define preset template paths
+preset_template = [
```
Comment (yingdachen) [inline app.py:478]: 如同线下讨论，不要直接展示文件名

另外要整体de-brand paiya

File: facechain/constants.py [hunk 10]
```diff
@@ -23,4 +23,5 @@
      'add_prompt_style': 'red, hanfu, tiara, crown, '},
 ]
 
-
+paiya_default_positive = 'beautiful, cool, finely detail, light smile, extremely detailed CG unity 8k wallpaper, huge filesize, best quality, realistic, photo-realistic, ultra high res, raw phot, put on makeup'
```
Comment (yingdachen) [inline facechain/constants.py:26]: 代码里不保留paiya的概念了

**Review by yingdachen — COMMENTED** [review]
很多从外部copy进来的代码，有可能用import吗？而不是copy code

File: facechain/train_text_to_image_paiya.py (hunk 5 above)
Comment (bubbliiiing) [reply facechain/train_text_to_image_paiya.py:1]: delete now

File: facechain/train_text_to_image_paiya.py (hunk 6 above)
Comment (bubbliiiing) [reply facechain/train_text_to_image_paiya.py:135]: i will be merged in preprocessing.py

File: app.py [hunk 11]
```diff
@@ -167,42 +255,48 @@ def run(
             uuid: str,
             instance_images: list,
     ) -> str:
-
+        # 检查CUDA是否可用
```
[the remark this answers is not in the document]
Comment (wuziheng) [reply app.py:255]: OK, done

File: app.py [hunk 12]
```diff
@@ -167,42 +255,48 @@ def run(
...
         if not torch.cuda.is_available():
-            raise gr.Error('CUDA is not available.')
+            raise gr.Error('CUDA不可用。')
```
[the remark this answers is not in the document]
Comment (wuziheng) [reply app.py:257]: OK, done

File: app.py [hunk 13]
```diff
@@ -167,42 +255,48 @@ def run(
...
+
+        # 检查是否上传了图片
```
[the remark this answers is not in the document]
Comment (wuziheng) [reply app.py:259]: Done, this is a interesting bug.

File: app.py [hunk 14]
```diff
@@ -167,42 +255,48 @@ def run(
...
         train_lora_fn(foundation_model_path='ly261666/cv_portrait_model',
-                      revision='v2.0',
-                      output_img_dir=instance_data_dir,
-                      work_dir=work_dir)
+                    revision='v2.0',
+                    output_img_dir=instance_data_dir,
+                    work_dir=work_dir)
 
-        message = f'训练已经完成！请切换至 [形象体验] 标签体验模型效果(Training done, please switch to the inference tab to generate photos.)'
+        message = '训练完成！切换到 [形象体验] 标签体验模型效果。'
```
[the remark this answers is not in the document]
Comment (wuziheng) [reply app.py:296]: Will done now~

File: app.py (hunk 9 above)
Comment (wuziheng) [reply app.py:475]: OK

Comment (wuziheng) [reply app.py:475]: 我已经修改成glob(dir)的形式并 de-brand， 
但我个人觉得， glob(dir) 相比直接写出文件名，不便于修改和debug。
所以我任然保留了原有的presets=[xxx.jpg] 的注释

File: facechain/constants.py (hunk 10 above)
Comment (wuziheng) [reply facechain/constants.py:26]: done, fix to inpaint

Comment (wuziheng) [reply facechain/constants.py:26]: 1. 统一删除了变量名中的paiya， replace with inpaint
2. 注释中的PAIYA 统一酌情使用Inpaint/Ensemble/PR104 代指，用于指代部分技术名称和link

File: app.py (hunk 8 above)
Comment (bubbliiiing) [reply app.py]: 已经修改为snapshot_download

File: facechain/kohya_lora.py [hunk 15]
```diff
@@ -0,0 +1,1310 @@
...
+        if self.rank_dropout is not None and self.training:
+            mask = torch.rand((lx.size(0), self.lora_dim), device=lx.device) > self.rank_dropout
+            if len(lx.size()) == 3:
+                mask = mask.unsqueeze(1)  # for Text Encoder
+            elif len(lx.size()) == 4:
+                mask = mask.unsqueeze(-1).unsqueeze(-1)  # for Conv2d
+            lx = lx * mask
+
+            # scaling for rank dropout: treat as if the rank is changed
+            # maskから計算することも考えられるが、augmentation的な効果を期待してrank_dropoutを用いる
```
[the remark this answers is not in the document]
Comment (bubbliiiing) [reply facechain/kohya_lora.py]: 已经删除，此前为kohya相关训练技术，切换到了facechain原有的lora训练代码

**Comment by bubbliiiing** [comment]
两张照片训练：
image

inference结果：

File: facechain/data_process/preprocessing.py [hunk 16]
```diff
@@ -1,19 +1,72 @@
...
+            if ('hair' in tag or 'face' in tag or 'mouth' in tag or 'skin' in tag or 'smile' in tag):
+                if not tag in add_prompt_style:
+                    add_prompt_style.append(tag)
+
+    if len(add_prompt_style) > 0:
+        add_prompt_style = ", ".join(add_prompt_style) + ', '
+    else:
+        add_prompt_style = ''
+    validation_prompt = trigger_style + add_prompt_style
+    return validation_prompt, neg_prompt
```
Comment (sunbaigui) [inline facechain/data_process/preprocessing.py:69]: 这部分是否可以和目前主干代码做融合，减少代码行数。看看是否有耦合的必要，还是说长期来看解耦更好。

File: README.md [hunk 17]
```diff
@@ -42,6 +43,11 @@ FaceChain is powered by [ModelScope](https://github.com/modelscope/modelscope).
 
 
 # Features
+- Support a series of new tricks to stableize Lora training
```
Comment (yingdachen) [inline README.md:46]: stablize

Introduce beta version of Inpaint, with techniques such as in-trianing validation and post-training ensemble to stablize LoRA.

File: README.md [hunk 18]
```diff
@@ -42,6 +43,11 @@ FaceChain is powered by [ModelScope](https://github.com/modelscope/modelscope).
...
+  - Description
+    - Add faceid evaluate in train_text_to_image_lora, which will generate validation image during training
+    - Add ensemble after Lora training according to faceid scores with training datasets
+    - Add InpaintTab, which not display in gradio,  Use Lora and template to generate artistic photo
```
Comment (yingdachen) [inline README.md:50]: Add Inpaint Tab (currently in beta and not displayed by default) in Gradio, which generates artistic template-based images.

File: README_ZH.md [hunk 19]
```diff
@@ -21,6 +21,7 @@ FaceChain的模型由[ModelScope](https://github.com/modelscope/modelscope)开
 
 
 # News
+- 为Lora训练添加验证和根据face_id的融合，并添加InpaintTab（目前隐藏在gradio中）. (2023-08-28)
```
Comment (yingdachen) [inline README_ZH.md:24]: 目前在Gradio界面上暂时默认隐藏

File: facechain/data_process/face_process_utils.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+        retinaface_box[2] = np.clip(retinaface_box[2] + face_width * (crop_ratio - 1) / 2, 0, w - 1)
+        retinaface_box[3] = np.clip(retinaface_box[3] + face_height * (crop_ratio - 1) / 2, 0, h - 1)
+        retinaface_box = np.array(retinaface_box, np.int32)
+
+        # Detect keypoints
+        retinaface_keypoints = np.reshape(retinaface_result['keypoints'][0], [5, 2])
+        retinaface_keypoints = np.array(retinaface_keypoints, np.float32)
+
+        # Mask part
+        # retinaface_crop = Image.fromarray(image).crop(tuple(np.int32(retinaface_box)))
```
Comment (yingdachen) [inline facechain/data_process/face_process_utils.py:40]: remove

File: facechain/data_process/preprocessing.py [hunk 21]
```diff
@@ -197,6 +250,64 @@ def get_mask_head(result):
...
+            embedding   = np.array(face_recognition(sub_image)[OutputKeys.IMG_EMBEDDING])
+            score       = face_quality_func(sub_image)[OutputKeys.SCORES]
+            score       = 0 if score is None else score[0]
+
+            face_id_scores.append(embedding)
+            quality_scores.append(score)
+            face_angles.append(angle)
+            selected_paths.append(jpg)
+        except:
+            print(jpg, "is error skip")
```
Comment (yingdachen) [inline facechain/data_process/preprocessing.py:303]: excpet e: 
skipping {jpg} due to processing error {e}

File: facechain/data_process/preprocessing.py [hunk 22]
```diff
@@ -210,20 +321,39 @@ def __init__(self):
                                                  'damo/cv_resnet34_face-attribute-recognition_fairface', model_revision='v2.0.2')
         self.facial_landmark_confidence_func = pipeline(Tasks.face_2d_keypoints,
                                                         'damo/cv_manual_facial-landmark-confidence_flcm', model_revision='v2.5')
+        # embedding
+        self.face_recognition = pipeline(Tasks.face_recognition, model='damo/cv_ir101_facerecognition_cfglint')
```
Comment (yingdachen) [inline facechain/data_process/preprocessing.py:325]: 加上model_revision吧。确保稳定性。

下同

File: app.py [hunk 23]
```diff
@@ -176,6 +209,63 @@ def launch_pipeline(uuid,
     else:
         yield ["生成失败，请重试(Generating failed, please retry)！", outputs_RGB]
```
Comment (yingdachen) [inline app.py:210]: Generation，顺手改一下

File: app.py [hunk 24]
```diff
@@ -176,6 +209,63 @@ def launch_pipeline(uuid,
...
+    instance_data_dir = os.path.join('/tmp', uuid, 'training_data', output_model_name)
+
+    # we use ensemble model, if not exists fallback to original lora
+    lora_model_path = f'/tmp/{uuid}/{output_model_name}/ensemble/'
+    if not os.path.exists(lora_model_path):
+        lora_model_path = f'/tmp/{uuid}/{output_model_name}/'
+
+    gen_portrait_inpaint = GenPortraitInpaint(crop_template=False, short_side_resize=512)
+    
+    cache_model_dir = snapshot_download("bubbliiiing/controlnet_helper", revision="v2.2") # '/mnt/zhoulou.wzh/AIGC/model_data/'
```
Comment (yingdachen) [inline app.py:241]: remove comment。# '/mnt/zhoulou.wzh/AIGC/model_data/'

File: app.py [hunk 25]
```diff
@@ -176,6 +209,63 @@ def launch_pipeline(uuid,
...
+    outputs = future.result()
+    outputs_RGB = []
+    for out_tmp in outputs:
+        outputs_RGB.append(cv2.cvtColor(out_tmp, cv2.COLOR_BGR2RGB))
+    image_path = './lora_result.png'
+    if len(outputs) > 0:
+        result = concatenate_images(outputs)
+        cv2.imwrite(image_path, result)
+
+        yield ["生成完毕(Generating done)！", outputs_RGB]
```
Comment (yingdachen) [inline app.py:265]: Generation，下同

File: app.py [hunk 26]
```diff
@@ -184,22 +274,30 @@ def __init__(self):
...
+
+        # Check Instance Valid
         if instance_images is None:
             raise gr.Error('您需要上传训练图片(Please upload photos)！')
-        if len(instance_images) > 10:
-            raise gr.Error('您需要上传小于10张训练图片(Please upload at most 10 photos)！')
+
+        # Limit input Image
+        if len(instance_images) > 20:
+            raise gr.Error('请最多上传20张训练图片(20 images most!)')
```
Comment (yingdachen) [inline app.py:291]: at most

File: app.py [hunk 27]
```diff
@@ -369,12 +484,110 @@ def inference_input():
     return demo
 
 
+# Define preset template paths
```
Comment (yingdachen) [inline app.py:487]: remove

File: app.py [hunk 28]
```diff
@@ -369,12 +484,110 @@ def inference_input():
...
+                    choices=HOT_MODELS,
+                    type="value",
+                    value=HOT_MODELS[0]
+                )
+                
+                template_gallery_list = [(i, i) for i in preset_template]
+                gallery = gr.Gallery(template_gallery_list).style(grid=4, height=300)
+                
+                # original select version
+                # selected_template_images = gr.CheckboxGroup(choices=preset_template, label="Select Preset Templates")
```
Comment (yingdachen) [inline app.py:518]: remove

File: facechain/data_process/preprocessing.py (hunk 21 above)
Comment (wuziheng) [reply facechain/data_process/preprocessing.py:303]: done
fix to
except Exception as e:
    print(f'skipping {jpg} due to processing error {e}')

File: README.md (hunk 18 above)
Comment (wuziheng) [reply README.md:50]: DONE

**Review by sunbaigui — APPROVED** [review]

**Review by yingdachen — APPROVED** [review]

**Review by yingdachen — APPROVED** [review]

**Pull request merged**